docs/fault_tolerance.rst

```diff
@@ -0,0 +1,1522 @@
+its own atomicity guarantees despite the fact that collectives are not atomic.
+Specifically, the body of a ``with live_devices`` block is guaranteed to either
+complete successfully on all processes or raise an exception on all processes.
+More concretely, if we consider the code snippet below, either every process
+executes branch A or every process executes branch B. It is impossible for some
+processes to execute A while others execute B.
+
+.. code-block:: python
+
+    try:
+      with live_devices(jax.live_devices()) as devices:
+        ...
+    except Exception as e:
+      ... # Branch A
+    else:
+      ... # Branch B
+
+.. warning::
+
+    A ``with live_devices`` block does not guarantee atomicity if the code
+    block non-deterministically raises exceptions for reasons other than
+    collectives that fail because of a crashed process. For example, if one
+    process raises an exception because it runs out of memory, this exception
+    will not be propagated to the other processes.
+
+Recall that JAX uses `asynchronous dispatch`_. Operations like ``jnp.sum`` do
+not block until the operation is complete. Instead, they return ``jax.Arrays``
+that act as futures. This asynchrony can interact with ``live_devices`` in
+unexpected ways. For example, consider the following code that performs a
+``jnp.sum``, assigns the result to ``y``, and then prints ``y``:
+
+.. code-block:: python
+
+    x = ...
+    y = ...
+    try:
+      with live_devices(jax.live_devices()) as devices:
+        y = jnp.sum(x)
+    except Exception as e:
+      ... # Branch A
+    else:
+      ... # Branch B
+    print(y)
+
+Imagine that the ``with live_devices`` block executes successfully on all
+processes. That is, all processes execute branch B. This only guarantees that
+every process successfully created a future and assigned it to ``y``. The
+actual computation of the ``jnp.sum`` may be delayed until outside the block.
+Thus, some processes might successfully complete the ``jnp.sum`` and print the
+value of ``y`` while other processes fail to complete the ``jnp.sum`` and raise
+an exception when trying to print ``y``.
+
+To avoid this, use ``jax.block_until_ready`` to ensure that computations are
+performed within the ``with live_devices`` block. The code snippet below, which
+now calls ``jax.block_until_ready`` when assigning to ``y``, guarantees that
+every process will successfully execute the ``jnp.sum`` or every process will
+raise an exception.
+
+.. code-block:: python
+
+    x = ...
+    y = ...
+    try:
+      with live_devices(jax.live_devices()) as devices:
+        y = jax.block_until_ready(jnp.sum(x))
+    except Exception as e:
+      ... # Branch A
+    else:
+      ... # Branch B
+    print(y)
+
+See :ref:`part3` for details on how atomicity is implemented.
+
+Sharding Arrays
+^^^^^^^^^^^^^^^
+
+TODO: Double check if ``jax.device_put`` works when some machines have
```

Comment:
Not sure if you want to address this before merging. I'm curious.

Reply:
I'll leave it for later.